File: leetcode/RangeSumQuery2D-Mutable/a.cpp
/* Copyright (C) 2019 by iamslash */

#include <cstdio>
#include <vector>

// 20ms 92.99% 14.6MB 21.31%
// sumRegion: O(W) O(HW)
//    update: O(W) O(HW)
class NumMatrix {
 private:
  std::vector<std::vector<int>> m_rowsums;
 public:
  NumMatrix(std::vector<std::vector<int>>& M) {
    if (M.empty() || M[0].empty())
      return;
    int h = M.size();
    int w = M[0].size();
    for (int y = 0; y < h; ++y) {
      std::vector<int> rowsum;
      int sum = 0;
      for (int x = 0; x < w; ++x) {
        sum += M[y][x];
        rowsum.push_back(sum);
      }
      m_rowsums.push_back(rowsum);
    }
  }
    
  void update(int r, int c, int val) {
    int leftsum = c > 0 ? m_rowsums[r][c-1] : 0;
    int delta = val - (m_rowsums[r][c] - leftsum);
    for (int x = c; x < m_rowsums[r].size(); ++x)
      m_rowsums[r][x] += delta;
  }
    
  int sumRegion(int lty, int ltx, int rby, int rbx) {
    int sum = 0;
    for (int y = lty; y <= rby; ++y) {
      int leftsum = ltx > 0 ? m_rowsums[y][ltx-1] : 0;
      sum += m_rowsums[y][rbx] - leftsum;
    }
    return sum;      
  }
};

int main() {
  
  // std::vector<std::vector<int>> M = {
  //   {3, 0, 1, 4, 2},
  //   {5, 6, 3, 2, 1},
  //   {1, 2, 0, 1, 5},
  //   {4, 1, 0, 1, 7},
  //   {1, 0, 3, 0, 5}
  // };

  std::vector<std::vector<int>> M = {
    {1},
    {2}
  };

  NumMatrix nm(M);
  printf("%d\n", nm.sumRegion(0, 0, 0, 0));
  printf("%d\n", nm.sumRegion(1, 0, 1, 0));
  printf("%d\n", nm.sumRegion(0, 0, 1, 0));
  nm.update(0, 0, 3);
  nm.update(1, 0, 5);
  printf("%d\n", nm.sumRegion(0, 0, 1, 0));

  return 0;
}

File: leetcode/RangeSumQuery2D-Mutable/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2021 by iamslash

# 3 0 1 4 2    3  3  4  7  9
# 5 6 3 2 1    5 11 14 16 17
# 1 2 0 1 5    1  3  3  4  9
# 4 1 0 1 7    4  5  5  6 13
# 1 0 3 0 5    1  1  4  4  9

from typing import List

# 128ms 91.05% 16.9MB 32.38%
# partial sum
# sumRegion: O(W)
#    update: O(W)
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        for line in matrix:
            for x in range(1, len(line)):
                line[x] += line[x-1]
        self.matrix = matrix

    def update(self, row: int, col: int, val: int) -> None:
        w = len(self.matrix[0])
        org = self.matrix[row][col]
        if col > 0:
            org -= self.matrix[row][col-1]
        diff = org - val
        for y in range(col, w):
            self.matrix[row][y] -= diff

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        sum = 0
        for x in range(row1, row2+1):
            sum += self.matrix[x][col2]
            if col1 > 0:
                sum -= self.matrix[x][col1-1]
        return sum
        
if __name__ == "__main__":
    nm = NumMatrix([
        [3, 0, 1, 4, 2],
        [5, 6, 3, 2, 1],
        [1, 2, 0, 1, 5],
        [4, 1, 0, 1, 7],
        [1, 0, 3, 0, 5]])
    # 8
    print(nm.sumRegion(2, 1, 4, 3))
    nm.update(3, 2, 2)
    # 10
    print(nm.sumRegion(2, 1, 4, 3))
    
